resources/Microsoft.Windows.Assertion/Microsoft.Windows.Assertion.psm1

```diff
@@ -0,0 +1,461 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version Latest
+
+enum PnPDeviceState
+{
+    OK
+    ERROR
+    DEGRADED
+    UNKNOWN
+}
+
+[DSCResource()]
+class OsEditionId
+{
+    # Key required. Do not set.
+    [DscProperty(Key)]
+    [string]$SID
+
+    [DscProperty(Mandatory)]
+    [string] $RequiredEdition
+
+    [DscProperty(NotConfigurable)]
+    [string] $Edition
+
+    [OsEditionId] Get()
+    {
+        $this.Edition = Get-ComputerInfo | Select-Object -ExpandProperty WindowsEditionId
+
+        return @{
+            RequiredEdition = $this.RequiredEdition
+            Edition         = $this.Edition
+        }
+    }
+
+    [bool] Test()
+    {
+        $currentState = $this.Get()
+        return $currentState.Edition -eq $currentState.RequiredEdition
+    }
+
+    [void] Set()
+    {
+        # This resource is only for asserting the Edition ID requirement.
```

Comment:
I am wondering if it makes sense to throw an error if this is ever called. That way the user doesn't accidentally think that something was actually set.

Reply:
I'm not sure. 

If I'm not mistaken, the way `winget configure` works is that it will test the resource, if it is not in the desired state, it will attempt to set it, then it checks to be sure that it set properly. If setting throws an error, I assume WinGet would be able to catch it?

Reply:
It looks like I was mistaken, and that WinGet doesn't actually check that the configuration was applied successfully before moving on. In order for these to work as intended, each resource will have to `throw` when the set method is called

Reply:
I've now added the functionality for it to throw an error when the setter is called along with related tests. Here's some sample output : 

```powershell
PS C:\Users\Trenly> Invoke-DscResource -Name PnPDevice -ModuleName Microsoft.Windows.Assertion -Method Set -Property @{Ensure='Absent';Status='OK'}
OperationStopped: Assertion Failed. One or more PnP devices found which matched the parameters
PS C:\Users\Trenly> Invoke-DscResource -Name PnPDevice -ModuleName Microsoft.Windows.Assertion -Method Set -Property @{Ensure='Present';FriendlyName='Test'}
OperationStopped: Assertion Failed. No PnP devices found which matched the parameters
PS C:\Users\Trenly> Invoke-DscResource -Name OsVersion -ModuleName Microsoft.Windows.Assertion -Method Set -Property @{MinVersion='12.01'}
OperationStopped: Assertion Failed. '10.0.22631' >= '12.01' evaluated to 'False'
PS C:\Users\Trenly> Invoke-DscResource -Name OsInstallDate -ModuleName Microsoft.Windows.Assertion -Method Set -Property @{Before='11/5/2024';After='11/3/2024'}
OperationStopped: Assertion Failed. '11/5/2024' >= '08/05/2024 09:22:31' >= '11/3/2024' evaluated to 'False'
```

And from WinGet -
```
Apply :: OsEditionId [os-edition]
  The configuration unit failed while attempting to apply the desired state.
Assertion Failed. Expected 'Enterprise Two' but received 'Enterprise'
<See the log file for additional details>
```